charms/knative-eventing/tests/unit/test_charm.py

```diff
@@ -1,67 +1,64 @@
 # Copyright 2022 Canonical Ltd.
 # See LICENSE file for licensing details.
-#
-# Learn more about testing at: https://juju.is/docs/sdk/testing
 
-import unittest
-from unittest.mock import Mock
+import pytest
+from k8s_resource_handler.exceptions import ErrorWithStatus
+from k8s_resource_handler.lightkube.mocking import FakeApiError
+from ops.model import ActiveStatus, BlockedStatus
 
-from ops.model import ActiveStatus
-from ops.testing import Harness
 
-from charm import OperatorTemplateCharm
+def test_events(harness, mocked_lightkube_client, mocker):
+    # Test install and config_changed event handlers are called
+    harness.begin()
+    install = mocker.patch("charm.KnativeEventingCharm._on_install")
+    config_changed = mocker.patch("charm.KnativeEventingCharm._on_config_changed")
 
+    harness.charm.on.install.emit()
+    install.assert_called_once()
+    install.reset_mock()
 
-class TestCharm(unittest.TestCase):
-    def setUp(self):
-        self.harness = Harness(OperatorTemplateCharm)
-        self.addCleanup(self.harness.cleanup)
-        self.harness.begin()
+    harness.charm.on.config_changed.emit()
+    config_changed.assert_called_once()
+    config_changed.reset_mock()
 
-    def test_config_changed(self):
-        self.assertEqual(list(self.harness.charm._stored.things), [])
-        self.harness.update_config({"thing": "foo"})
-        self.assertEqual(list(self.harness.charm._stored.things), ["foo"])
 
-    def test_action(self):
-        # the harness doesn't (yet!) help much with actions themselves
-        action_event = Mock(params={"fail": ""})
-        self.harness.charm._on_fortune_action(action_event)
+def test_on_install_active(harness, mocked_lightkube_client, mocker):
+    mocked_apply = mocker.patch(
+        "k8s_resource_handler.kubernetes._kubernetes_resource_handler.KubernetesResourceHandler.apply"
+    )
+    mocked_apply.return_value = None
 
-        self.assertTrue(action_event.set_results.called)
+    harness.begin()
+    harness.charm.on.install.emit()
```

Comment:
Nothing wrong with this here, but imo it feels cleaner to do `harness.charm._on_install("some fake event")` because I feel like you've already tested (`emit install --> handler fires`) above, and then here you're testing whether on_install actually results in `ActiveStatus`.  But there's nothing wrong with the present way, and doing it the alternate way does beg the question "then why do I have this harness at all?"